src/java.base/share/classes/java/lang/Double.java

```diff
@@ -707,51 +709,76 @@ public static String toHexString(double d) {
             // For infinity and NaN, use the decimal output.
             return Double.toString(d);
         else {
-            // Initialized to maximum size of output.
-            StringBuilder answer = new StringBuilder(24);
-
-            if (Math.copySign(1.0, d) == -1.0)    // value is negative,
-                answer.append("-");                  // so append sign info
-
-            answer.append("0x");
-
+            boolean negative = Math.copySign(1.0, d) == -1.0;
             d = Math.abs(d);
 
-            if(d == 0.0) {
-                answer.append("0.0p0");
+            if (d == 0.0) {
+                return negative ? "-0x0.0p0" : "0x0.0p0";
             } else {
+                // Check if the value is subnormal (less than the smallest normal value)
                 boolean subnormal = (d < Double.MIN_NORMAL);
 
                 // Isolate significand bits and OR in a high-order bit
-                // so that the string representation has a known
-                // length.
+                // so that the string representation has a known length.
+                // This ensures we always have 13 hex digits to work with (52 bits / 4 bits per hex digit)
                 long signifBits = (Double.doubleToLongBits(d)
-                                   & DoubleConsts.SIGNIF_BIT_MASK) |
-                    0x1000000000000000L;
+                        & DoubleConsts.SIGNIF_BIT_MASK) |
+                        0x1000000000000000L;
```

Comment:
As a small syntax improvement, I think using underscores between groups of four digits would help how the count is correct (that syntax was legal when this code was written. Something like

 0x1000_0000_0000_0000L